### DISCUSSION TITLE
does it work with C++ code?

### DISCUSSION THREAD

**Discussion by gfim** [danielpalme/ReportGenerator#536] [Q&A]
I've been trying to get ReportGenerator to report on our C++ codebase. My input is a Cobertura data file produced by CPPCoverage (https://github.com/atlaste/CPPCoverage). I run that through ReportGenerator to produce HTML. I get a directory containing HTML output but it's essentially empty. It has a Summary section with all counts zero. And the Coverage section says simply "No assemblies have been covered". I've also tried using CPPCoverage's native file format with exactly the same results.

Am I doing something wrong or is this tool not capable of producing output for C++ code?

Thanks for any input on this.

**Comment by danielpalme** [comment]
_ReportGenerator_ is agnostic about the used programming language.
I guess the Coberatura file generated by CppCoverage has some special formatting.
If you could share a sample file, I can have a closer look.

**Reply by gfim** [reply]
Thanks, Daniel. I couldn't see any reason why C++ wouldn't work, but I wasn't sure. I'll email you a Cobertura file.

Thanks,
  Graham

**Reply by gfim** [reply]
Ok, I added the -verbosity:verbose parameter and I can see what's happening - it's using the mprof parser instead of the Cobertura one. 
I modified the source code to force the Cobertura parser and it now works correctly.

In order to diagnose the problem, the first few lines of the Cobertura file are:

```
<?xml version="1.0" encoding="utf-8"?>
<coverage line-rate="0.120511">
<packages>
<package name="Program.exe" line-rate="0.120511">
<classes>
<class name="C:/git/Application1/target_hmi/src/ui/spx/managers/radiocallactivity.h" filename="/git/Application1/target_hmi/src/ui/spx/managers/radiocallactivity.h" line-rate="0">
<lines>
<line number="128" hits="0"/>
<line number="129" hits="0"/>
<line number="130" hits="0"/>
<line number="131" hits="0"/>
</lines>
```

**Reply by danielpalme** [reply]
Thanks for you investigation.
I just improved the the Cobertura/MBProf in de21e67291f7a6c42e5189577d1e674217751dbe.

The next release will contain this improvement, so you won't have to modify the coverage file manually any more.

**Reply by gfim** [reply]
Thank you Daniel - it's working correctly now.

And thank you for ReportGenerator - very useful tool!

Graham

**Reply by johndaintree** [reply]
I'm having what  I think is a similar problem, with native "C" code (NOT C++). 

I have an application which is a combination of C, C++, C++ CLR, and C#.

 I've used "Microsoft.CodeCoverage.Console.exe instrument" to instrument the native binaries.
 I've then  used  "Microsoft.CodeCoverage.Console.exe collect" to get coverage data for those binaries
 The resulting .coverage file, in Visual Studio, shows code coverage for the native C code (in a "Global Functions" section)
 
I've then used "CodeCoverage.exe analyze /output...." to convert the .coverage file to a .coveragexml file
Looking at the resulting XML appears to show that the native data is still there (I can see the .c filenames and the names of functions that are called)

I then do "reportgenerator -reports:./my.exe.coveragexml -targetdir:./coverage -reporttypes:html
do generate the HTML. 

In the html I can see coverage information (with source code, and coloring) for the C++ code and for the managed code. But I can't see any information for the native C code.

I can send the .coverage and .coveragexml files if it'd help.

Is native C code supported?

Thanks,
John.

**Reply by danielpalme** [reply]
@johndaintree 
As said above, _ReportGenerator_ is agnostic about the used programming language.

It would be great if you could share the `*.coveragexml` file(s) with me, then I can have a closer look.

**Reply by danielpalme** [reply]
@johndaintree 
Thanks for providing the coverage files.

The Problem is in the ".coverage.xml" file.
It contains `<function>` elements with an empty "type_name" attribute.
Therefore all those functions appear in the HTML report as a single empty element. 

I made a change in _ReportGenerator 5.1.16_ to fix the problem.
Instead of the empty name, the first `<range>` element within those functions is searched.
The file name of the corresponding file is then applied to the empty „type_name“ attribute (based on the „source_id“ attribute of the `<range>` element).

**Reply by johndaintree** [reply]
Wonderful, thanks for that, it is much better. Can I ask for a small tweak, and have the file extension included in the HTML? 

Many thanks again.

**Reply by danielpalme** [reply]
Thanks for your feedback (and for buying a PRO license).
I will have a look within the next days.

**Reply by danielpalme** [reply]
@johndaintree 
The file extension is now displayed (only for files that would be missing).
Release 5.1.17 will be available soon.
